### PR TITLE
DAOS-2265 build: Add GO build-ids

### PR DESCRIPTION
**Pull request by brianjmurrell**
Thanks to a hack by Michael MacDonald, we can now give GO binaries a
build-id.

Signed-off-by: Brian J. Murrell <brian.murrell@intel.com>

### CODE REVIEW

**Comment by brianjmurrell** [comment]
Functional failure is [DAOS-3945](https://jira.hpdd.intel.com/browse/DAOS-3945)

File: src/control/SConscript [hunk 1]
```diff
@@ -66,9 +66,27 @@ def install_go_bin(denv, gosrc, gopath, libs, name, install_name):
...
+            # into a 'unicode' type
+            return '0x' + buildid
+        else:
+            # Python 3 is a bytes, decode() it to a str
+            return '0x' + buildid.decode()
+    def go_ldflags():
+        return ' '.join(['-ldflags',
+                         '"-X main.daosVersion=%s' % DAOS_VERSION,
+                         '-B %s"' % gen_build_id(),
+                         '-v'])
```
Comment (mjmac) [inline src/control/SConscript]: Technically, `-v` is a flag for `go install`, but I don't think you need to rework it for that.

File: src/control/SConscript [hunk 2]
```diff
@@ -66,9 +66,27 @@ def install_go_bin(denv, gosrc, gopath, libs, name, install_name):
...
     if libs is not None:
         src.extend(libs)
+    # generate a unique build id per binary for use by RPM
+    # https://fedoraproject.org/wiki/PackagingDrafts/Go#Build_ID
+    def gen_build_id():
+        from os import urandom
+        from binascii import b2a_hex
+        buildid = b2a_hex(urandom(20))
+        if isinstance(buildid, str):
+            # Python 2 is a str already, using decode() here transforms it
```
Comment (mjmac) [inline src/control/SConscript]: Nit: Comment doesn't seem to match code anymore...

File: src/control/SConscript (hunk 1 above)
Comment (brianjmurrell) [reply src/control/SConscript]: Ah, right.  Of course.

File: src/control/SConscript (hunk 2 above)
Comment (brianjmurrell) [reply src/control/SConscript]: Not sure I'm seeing the mismatch.  The comment is basically saying, "use it as is" on Python 2 which is what the code is doing.

Comment (mjmac) [reply src/control/SConscript]: The comment in the python2 branch states:
"Python 2 is a str already, using decode() here transforms it into a 'unicode' type" ... But in the python2 branch, there's no use of decode. Was just something that jumped out at me as a reader. Something like "Python 2 is a str, no need to decode" would be clearer.

Comment (brianjmurrell) [reply src/control/SConscript]: Ahh.  I see the confusion.  If I were more verbose with my comment I guess I would have written:

_In Python 2 **buildid** is a str already; using decode() on it here transforms it into a 'unicode' type, so we're not doing that, like we do below for Python3 where we need to decode() it from a bytes into a str_

**Pull request merged**